tests/test_utils.sh

```diff
@@ -710,6 +710,51 @@ function test_is_function {
   unalias not_my_function
 }
 
+function test_hash_color {
+    PS1="$ "
+    lp_activate --no-config # For having _lp_foreground
+    _lp_hash_color "Debug"
+    assertContains "$lp_hash_color" "Debug"
+    # FIXME How to test color?
+}
+
+function test_join {
+    _lp_join "_" "A" "B " " " " C" " D " "EE"
+    assertEquals "A_B _ _ C_ D _EE" "$lp_join"
+
+    _lp_join "-" "A"
+    assertEquals "A" "$lp_join"
+
+    _lp_join "+" ""
+    assertEquals "" "$lp_join"
+
+    typeset -a arr
+    arr=(1 2 3)
+    _lp_join "/" "${arr[@]}"
+    assertEquals "1/2/3" "$lp_join"
+}
+
+function test_grep_fields {
+    filename=$(mktemp)
+    echo -e "key1:value1\nkey2:value2" > "$filename"
+    _lp_grep_fields "$filename" ":" "key1" "key2"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+0]} "value1"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+1]} "value2"
+    assertEquals ${#lp_grep_fields[@]} 2
+
+    echo -e " key1 :NOPE\nkey1:value1\nNOPE:NOPE\n key1:NOPE\nkey1 :NOPE" > "$filename"
+    _lp_grep_fields "$filename" ":" "key1"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+0]} "value1"
+    assertEquals ${#lp_grep_fields[@]} 1
+
+    echo -e "key1=value1\nkey2=value=key2=?\nkey3==val" > "$filename"
+    _lp_grep_fields "$filename" "=" "key1" "key2" "key3"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+0]} "value1"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+1]} "value=key2=?"
+    assertEquals ${lp_grep_fields[_LP_FIRST_INDEX+2]} "=val"
+    assertEquals ${#lp_grep_fields[@]} 3
```

Comment:
Good test cases, I would also like to see one where the keys are in a different order in the source file than when they are passed in as arguments. Should be as simple as adding a `_lp_grep_fields "$filename" "=" "key3" "key1" "key2"` then the exact same asserts.

Also needs a test case for when the key is NOT in the file. The index should be unset, so just verifying that. (can test for unset value with `[[ -z ${array[i]+x} ]]`, so probably like `assertTrue "found match for non existent key:value" '[[ -z "${lp_grep_fields[_LP_FIRST_INDEX+1]+x}" ]]'`

Reply:
Interestingly, switching to `printf` introduced a very subtle bug: the `while read line` construction will not read the last line if the read file does not end with an explicit new line. To ensure to always read the file up to the very last line, one should use `while IFS='' read -r line || [[ -n "$line" ]]`. This may affect your `_lp_python_env` function as well.

Reply:
Yes, that is a valid observation. There are a lot of utilities that assume all files end in a newline, but unless you can think of a drawback to assuming characters without a newline is a valid line, I say let's use the `while IFS='' read -r line || [[ -n "$line" ]]` form.

And yes, I was planning on converting `_lp_python_env()` to use this new search function once we had this merged.